_posts/2019-04-05-omakase-typescript.markdown

```diff
@@ -0,0 +1,171 @@
+say upfront what can be inside that variable, which is only strings, so:
+
+```js
+// Before in JS
+var name = "Danger"
+// Also OK
+var name = 1
+var name = false
+var name = { first: "Danger", last: "McMaslen" }
+
+// Before in Java
+String name = "Danger";
+// Not OK, the code wouldn't be accepted by Java
+String name = 1;
+
+// This example isn't quite what you'd write, but it shows the difference well
+Map<String, String> person = new HashMap<String, String>();
+person.put("first", "Danger");
+person.put("last", "McMaslen");
+String name = person;
+```
+
+These trade-offs make sense in the context for which these languages were built. JavaScript was **originally**
+designed in a a short time, to specifically do a small amount of things. Java on the other hand was built
+specifically to make very, very large amounts of easy to read with - which means sometimes you have to do more
+writing/reading chores. It's an implicit vs explicit thing.
+
+### What is TypeScript?
+
+TypeScript is a programming language - it contains all of JavaScript, and then a bit more. So in our case above,
+lets compare the scripts for "Hello, Danger" in JavaScript vs TypeScript:
+
+```js
+// JavaScript
+var name = "Danger"
+console.log("Hello, " + danger)
+
+// TypeScript
+var name = "Danger"
+console.log("Hello, " + danger)
+
+// Yep, you're not missing something, there's no difference
+```
+
+Due to TypeScript aiming to only _extend_ JavaScript, your normal JavaScript code should work fine with TypeScript.
+The parts of the JavaScript language which TypeScript extends are specifically about trying to make it easier to be
+more explicit in your code.
+
+```diff
+- var name = "Danger"
++ var name: string = "Danger"
+console.log("Hello, " + danger)
+```
+
+This allows the reader to be certain that `name` will only be a string. Annotating your variables also allows
+TypeScript to check this for you.
+
+Simply speaking, we call these annotation "Types" and this is the "Type" in TypeScript. The tag-line for TypeScript
+is "JavaScript which scales" which is a statement that these extra type annotations allows you to work on bigger
+projects. This is because you can verify how correct your code is, and as many people read and write it then you
+don't need to fit it all in your head.
+
+TypeScript extends JavaScript more, but this is the core concept - and that's enough for now.
+
+### Why does Artsy use TypeScript?
+
+Artsy definitely isn't the size of Microsoft! Artsy is about 30 engineers, and Microsoft are about 60,000. However,
+some of our problems are the same. The "front end" developers at Artsy build apps which are thousands of files big.
+Any single change to one file, could affect any other file. A good way to think about this is how in a story,
+changing a conversation earlier can trigger a butterfly effect of changes later down the line.
+
+Typically, the need to ensure all your changes do what you want is less of a problem for people building websites.
```

Comment:
> ensure all your changes do what you want

This might be easier to grok as

> ensure there are no bugs 

though not semantically identical

Reply:
Yeah, I think bugs is common enough that peeps  will get what I'm going for